[WICKET-7153] 1) make system mapper more reusable by making newResour…

reiern70 opened the pull request:
1) make system mapper more reusable by making newResourceReferenceMapper protected and 
2) roll a hash code version of ResourceReferenceMapper that encodes class names by using their hash code.]

See

https://issues.apache.org/jira/browse/WICKET-7153

martin-g commented on wicket-core/src/main/java/org/apache/wicket/core/request/mapper/HashBasedResourceReferenceMapper.java:
```diff
@@ -0,0 +1,121 @@
+ * </p>
+ *
+ * <p>
+ *     Caveat: we don't take into account has collisions. I.e. to to different class names having the same hash code.
```
```suggestion
 *     Caveat: we don't take into account hash collisions. I.e. to to different class names having the same hash code.
```

martin-g replied:
`I.e. to to different class` sounds wrong. Please re-phrase it.

martin-g commented on wicket-core/src/main/java/org/apache/wicket/core/request/mapper/HashBasedResourceReferenceMapper.java:
```diff
@@ -0,0 +1,121 @@
+ *    on your Application class.
+ * </p>
+ */
+public class HashBasedResourceReferenceMapper extends ParentPathReferenceRewriter
```
This class needs unit tests!

martin-g commented on wicket-core/src/main/java/org/apache/wicket/core/request/mapper/HashBasedResourceReferenceMapper.java:
```diff
@@ -0,0 +1,121 @@
+		protected String getClassName(Class<?> scope)
+		{
+			String name = super.getClassName(scope);
+			long hash = scope.hashCode();
```
The class javadoc says `that encodes class names as hash codes`, i.e. `String.hashCode()`, but it is actualy `Class.hashCode()`

Probably you want:
```suggestion
			long hash = name.hashCode();
```
?!

reiern70 commented:
@martin-g Thanks a lot for review and comments. I have addressed them

- I have added a test
- I re-worked the code so that users can use a different hash "generator".
